adapters/repos/db/replication.go

```diff
@@ -46,186 +46,182 @@ type Replicator interface {
-	shardName, requestID string, refs []objects.BatchReference,
+func (db *DB) ReplicateReferences(ctx context.Context, class,
+	shard, requestID string, refs []objects.BatchReference,
 ) replica.SimpleResponse {
-	index := db.GetIndex(schema.ClassName(indexName))
-	if index == nil {
-		return replica.SimpleResponse{
-			Errors: []string{fmt.Sprintf("local index %q not found", indexName)},
-		}
+	index, pr := db.replicatedIndex(class)
+	if pr != nil {
+		return *pr
 	}
 
-	return index.ReplicateReferences(ctx, shardName, requestID, refs)
+	return index.ReplicateReferences(ctx, shard, requestID, refs)
 }
 
-func (db *DB) CommitReplication(ctx context.Context, indexName,
-	shardName, requestID string,
+func (db *DB) CommitReplication(ctx context.Context, class,
+	shard, requestID string,
 ) interface{} {
-	index := db.GetIndex(schema.ClassName(indexName))
-	if index == nil {
-		return replica.SimpleResponse{
-			Errors: []string{fmt.Sprintf("local index %q not found", indexName)},
-		}
+	index, pr := db.replicatedIndex(class)
+	if pr != nil {
+		return *pr
 	}
 
-	return index.CommitReplication(ctx, shardName, requestID)
+	return index.CommitReplication(ctx, shard, requestID)
 }
 
-func (db *DB) AbortReplication(ctx context.Context, indexName,
-	shardName, requestID string,
+func (db *DB) AbortReplication(ctx context.Context, class,
+	shard, requestID string,
 ) interface{} {
-	index := db.GetIndex(schema.ClassName(indexName))
-	if index == nil {
-		return replica.SimpleResponse{
-			Errors: []string{fmt.Sprintf("local index %q not found", indexName)},
-		}
+	index, pr := db.replicatedIndex(class)
+	if pr != nil {
+		return *pr
 	}
 
-	return index.AbortReplication(ctx, shardName, requestID)
+	return index.AbortReplication(ctx, shard, requestID)
 }
 
-func (i *Index) ReplicateObject(ctx context.Context, shard, requestID string, object *storobj.Object) replica.SimpleResponse {
-	i.backupStateLock.RLock()
-	defer i.backupStateLock.RUnlock()
-	localShard, ok := i.Shards[shard]
+func (db *DB) replicatedIndex(name string) (idx *Index, resp *replica.SimpleResponse) {
+	if idx = db.GetIndex(schema.ClassName(name)); idx == nil {
+		return nil, &replica.SimpleResponse{Errors: []replica.Error{
+			*replica.NewError(replica.StatusClassNotFound, name),
+		}}
+	}
+	return
+}
+
+func (i *Index) writableShard(name string) (*Shard, *replica.SimpleResponse) {
+	localShard, ok := i.Shards[name]
 	if !ok {
-		return replica.SimpleResponse{Errors: []string{"shard not found"}}
+		return nil, &replica.SimpleResponse{Errors: []replica.Error{
+			{Code: replica.StatusShardNotFound, Msg: name},
+		}}
+	}
+	if localShard.isReadOnly() {
+		return nil, &replica.SimpleResponse{Errors: []replica.Error{{
+			Code: replica.StatusReadOnly, Msg: name,
+		}}}
+	}
+	return localShard, nil
+}
+
+func (i *Index) ReplicateObject(ctx context.Context, shard, requestID string, object *storobj.Object) replica.SimpleResponse {
+	localShard, pr := i.writableShard(shard)
```

Comment:
Why is this `SimpleResponse` variable called `pr`?

Reply:
pr is a local variable which stands for pointer to response. pr == nil  if localshard exists